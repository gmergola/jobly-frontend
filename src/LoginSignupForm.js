import React, { useState, useEffect, useContext } from 'react';
import JoblyApi from "./HelperApi";
import { useHistory } from "react-router-dom";
import TokenContext from "./TokenContext";
import Alert from './Alert';
import "./login.css";
import blueLines from './images/blue-lines.gif';

/**LoginSignupForm:  parent component of alert, child component of homepage,routes
 * renders forms, sends request, receives token or error based on response,
 * authenticates user or Alerts user to error, redirects to company page upon login
*/
function LoginSignupForm() {
  const history = useHistory();
  const initialData = {
    username: '',
    password: '',
    first_name: '',
    last_name: '',
    email: ''
  }
  const [hideLogin, setHideLogin] = useState('Hidden');
  const [hideSignup, setHideSignup] = useState('Hidden');
  const [formData, setFormData] = useState({ ...initialData });
  const [logginIn, setlogginIn] = useState(false);
  const [signingUp, setsigningUp] = useState(false);
  const [errorMessage, setErrorMessage] = useState([]);
  const { setToken } = useContext(TokenContext);


  /*registerOrLogin: invokes login/register requests (depending on which form is submitted)
    then resets submit click states, and hides error messages if exposed
  */
  useEffect(function registerOrLogin() {

    /*logIn: uses login request method, sets token state with response, saves token in
    localStorage, redirects to companies once logged in via history */
    async function logIn() {

      try {
        let response = await JoblyApi.login(formData);
        window.localStorage.setItem('token', response);
        window.localStorage.setItem('username', formData.username);
        setToken(response);
        history.push("/companies");
      }
      catch (err) {
        setErrorMessage(messages=>([
          ...messages, ...err
        ]));
        console.error(err);
      }


    }

    /*signup: uses register request method, sets token state with response, saves token in
    localStorage, redirects to companies once logged in via history */
    async function signUp() {

      try {
        let response = await JoblyApi.signup(formData);
        window.localStorage.setItem('token', response);
        window.localStorage.setItem('username', formData.username);
        setToken(response);
        history.push("/companies");
      } catch (err) {
        setErrorMessage(messages=>([
          ...messages, ...err
        ]));
        console.error(err);
      }

    }


    if (logginIn) {
      logIn();
      setlogginIn(false);
      setFormData(initialData);
      setErrorMessage([]);


    }
    else if (signingUp) {
      signUp();
      setsigningUp(false);
      setFormData(initialData);
      setErrorMessage([]);

    }


  }, [signingUp, logginIn, formData, initialData, setToken, history])


  function handleSubmitSignUp(evt) {
    evt.preventDefault();
    setsigningUp(true);
  }

  function handleSubmitLogin(evt) {
    evt.preventDefault();
    setlogginIn(true);
    console.log("login", localStorage);

  }

  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData(data => ({
      ...data,
      [name]: value
    }))
  }

  //toggles our register/login forms to view the appropriate form
  function handleRegisterButton() {
    setHideLogin("Hidden");
    setHideSignup("");
  }
  function handleLoginButton() {
    setHideLogin("");
    setHideSignup("Hidden");
  }

  return (
    <div className="Login-container">
      <div ><Alert errors={errorMessage} /></div>
      {!window.localStorage.getItem("token") ?
      <div>
      <button className="login-signup-btn btn btn-primary" onClick={handleLoginButton}>Login</button>
      <br />
      <button className="login-signup-btn btn btn-secondary" onClick={handleRegisterButton}>Register</button>
      </div>
      :
      <img src={blueLines} alt="lines"/>}
      <br />
      <br />
      <form className={`"loginForm" "form-group" ${hideLogin}`} onSubmit={handleSubmitLogin}>
        <div className="input-group">
          <div className="input-group-prepend">
            <span className="input-group-text">Username</span>
          </div>
          <input className="form-control" onChange={handleChange} value={formData.username} name="username"></input>
        </div>
        <br />
        <div className="input-group">
          <div className="input-group-prepend">
          <span className="input-group-text">Password</span>
          </div>
          <input className="form-control" onChange={handleChange} value={formData.password} name="password"></input>
        </div>
        <br />
        <button className="login-btn btn btn-primary" type="submit">Submit</button>
      </form>


      <form className={`"signUpForm" ${hideSignup}`} onSubmit={handleSubmitSignUp}>
        <div className="input-group">
          <div className="input-group-prepend">
            <span className="input-group-text">First and last name</span>
          </div>
          <input className="form-control" onChange={handleChange} value={formData.first_name} name="first_name"></input><br />
          <input className="form-control" onChange={handleChange} value={formData.last_name} name="last_name"></input><br />
        </div>
        <br />

        <div className="input-group">
          <div className="input-group-prepend">
            <span className="input-group-text">Email</span>
          </div>
          <input className="form-control" onChange={handleChange} value={formData.email} name="email"></input>
        </div>
        <br />

        <div className="input-group">
          <div className="input-group-prepend">
            <span className="input-group-text">Username</span>
          </div>
          <input className="form-control" onChange={handleChange} value={formData.username} name="username"></input>
        </div>
        <br />

        <div className="input-group">
          <div className="input-group-prepend">
            <span className="input-group-text">Password</span>
          </div>
          <input className="form-control" onChange={handleChange} value={formData.password} name="password"></input>
        </div>
        <button className="signup-btn btn btn-secondary" type="submit">Submit</button>
      </form>
    </div>
  )

}

export default LoginSignupForm;