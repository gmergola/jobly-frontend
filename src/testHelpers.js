import React from "react";
import TokenContext from "./TokenContext";

const demoToken = 'TOKEN';

const TokenProvider = ({ children, currentToken = demoToken }) => (
  <TokenContext.Provider value={{currentToken}}>
    {children}
  </TokenContext.Provider>
);

export { TokenProvider };
