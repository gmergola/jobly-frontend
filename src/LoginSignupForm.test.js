import React from "react";
import { render } from "@testing-library/react";
import LoginSignupForm from "./LoginSignupForm";
import { MemoryRouter } from "react-router";
import { TokenProvider } from "./testHelpers";


it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <TokenProvider>
        <LoginSignupForm />
      </TokenProvider>
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <TokenProvider>
        <LoginSignupForm />
      </TokenProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});