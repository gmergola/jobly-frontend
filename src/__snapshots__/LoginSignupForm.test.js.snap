// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Login-container"
  >
    <div />
    <div>
      <button
        class="login-signup-btn btn btn-primary"
      >
        Login
      </button>
      <br />
      <button
        class="login-signup-btn btn btn-secondary"
      >
        Register
      </button>
    </div>
    <br />
    <br />
    <form
      class="\\"loginForm\\" \\"form-group\\" Hidden"
    >
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="input-group-text"
          >
            Username
          </span>
        </div>
        <input
          class="form-control"
          name="username"
          value=""
        />
      </div>
      <br />
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="input-group-text"
          >
            Password
          </span>
        </div>
        <input
          class="form-control"
          name="password"
          value=""
        />
      </div>
      <br />
      <button
        class="login-btn btn btn-primary"
        type="submit"
      >
        Submit
      </button>
    </form>
    <form
      class="\\"signUpForm\\" Hidden"
    >
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="input-group-text"
          >
            First and last name
          </span>
        </div>
        <input
          class="form-control"
          name="first_name"
          value=""
        />
        <br />
        <input
          class="form-control"
          name="last_name"
          value=""
        />
        <br />
      </div>
      <br />
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="input-group-text"
          >
            Email
          </span>
        </div>
        <input
          class="form-control"
          name="email"
          value=""
        />
      </div>
      <br />
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="input-group-text"
          >
            Username
          </span>
        </div>
        <input
          class="form-control"
          name="username"
          value=""
        />
      </div>
      <br />
      <div
        class="input-group"
      >
        <div
          class="input-group-prepend"
        >
          <span
            class="input-group-text"
          >
            Password
          </span>
        </div>
        <input
          class="form-control"
          name="password"
          value=""
        />
      </div>
      <button
        class="signup-btn btn btn-secondary"
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
